Re-lands platform brightness support on iOS

matthew-carroll opened the pull request:
Re-lands platform brightness support on iOS with an #ifdef for CI.

matthew-carroll commented:
Weird that this seemed to compile for me with the wrong syntax...I'll switch to #else

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:
```diff
@@ -940,6 +957,24 @@ - (BOOL)isAlwaysUse24HourFormat {
+- (NSString*)brightnessMode {
+#if defined(__IPHONE_13_0)
+  if (@available(iOS 13, *)) {
+    UIUserInterfaceStyle style = UITraitCollection.currentTraitCollection.userInterfaceStyle;
```
Why aren't you just using `self.traitCollection` here? That would avoid the SDK problem, right?

matthew-carroll replied:
I'll give that a try

matthew-carroll commented:
@chinmaygarde @stuartmorgan I switched to the VC's traitCollection and removed the API check

chinmaygarde commented on shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.m:
```diff
@@ -23,4 +25,85 @@ - (void)testBinaryMessenger {
+  id settingsChannel = OCMClassMock([FlutterBasicMessageChannel class]);
+  OCMStub([engine settingsChannel]).andReturn(settingsChannel);
+
+  FlutterViewController* vc = [[FlutterViewController alloc] initWithEngine:engine
```
Are these tests running with ARC? Because these look like leaks. The rest of the engine does not use ARC.

matthew-carroll replied:
No idea. I was just continuing with what I already found in this file. If you have any specific recommendations for changes, I can make them, but I'm working from a lack of iOS knowledge here.

matthew-carroll replied:
@chinmaygarde @stuartmorgan do either of you have any specific instructions for what should be done here?

matthew-carroll replied:
@dnfield @gspencergoog do either of you know what changes should be made here to avoid leaks? I don't know what changes to make to bring this test code into compliance...

stuartmorgan replied:
I would turn on ARC for the test file, since that would be easy (it looks like the test project is doing it file-by-file, which seems like an odd approach; I would just enable it for the whole project).

Failing that, add an `autorelease` here.

matthew-carroll replied:
I don't know how to turn on ARC and I also don't know what the structure of our testing system is here so that sounds like a risky thing for me to do.

I'll add autorelease if that solves the problem. Do I add it only to the vc references?

matthew-carroll replied:
@LongCatIsLooong helped me get ARC enabled

dnfield replied:
We may want to have ARC off for the project so that we can explicitly test deallocating things...

matthew-carroll replied:
I believe I only enabled it for this file. Do you want me to change something?

dnfield replied:
No, that's fine - my reply was more directed towards @stuartmorgan's comment :)

stuartmorgan replied:
Well since the only test in that file is leaking the view controller, I'm pretty sure it's not deliberately off to test dealloc...

chinmaygarde commented on testing/ios/IosUnitTests/run_tests.sh:
```diff
@@ -12,6 +12,6 @@ fi
 set -o pipefail && xcodebuild -sdk iphonesimulator \
   -scheme IosUnitTests \
-  -destination 'platform=iOS Simulator,name=iPhone SE,OS=12.2' \
+  -destination 'platform=iOS Simulator,name=iPhone 8,OS=13.0' \
```
This simulator variant won't be available on ToT.

matthew-carroll replied:
Reverted the OS version, but kept it on iPhone 8 because my local environment doesn't even offer an iPhone SE

dnfield replied:
FWIW you should be able to add iPhone SE locally, but AFAIK know of our tests depend on the simulator type at this point.

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:
```diff
@@ -690,6 +695,11 @@ - (CGFloat)statusBarPadding {
 
+- (void)viewWillLayoutSubviews {
+  [super viewWillLayoutSubviews];
+  [self onUserSettingsChanged:nil];
```
Looking more broadly at the patch rather than just the part that was failing before: why are all these layout-based methods getting onUserSettingChanged: calls? traitCollectionDidChange: makes sense, but I don't understand why a relayout would change any settings (and especially why each relayout would need to update the settings 3 times, as this change appears to be doing.)

matthew-carroll replied:
Here's the guide I was following:
https://developer.apple.com/documentation/appkit/supporting_dark_mode_in_your_interface?language=objc

stuartmorgan replied:
That guide says that all four of those methods are called on appearance changes, and that appearance-sensitive code should be moved into one of them. That's very different than saying that you should re-run *all* appearance-sensitive code in *all* of them.

matthew-carroll replied:
Alright. Do you have a recommendation? That guide does not disambiguate execution paths that lead to those methods, so it's not clear to me which subset of those methods will actually account for all use-cases where platform brightness might change. Do you know which ones we want?

stuartmorgan replied:
The documentation as I read it is saying that all of those are called when the appearance changes. I would implement only `traitCollectionDidChange:` (which best matches the conceptual use here) until/unless you have evidence that there are cases where that isn't called.

matthew-carroll replied:
I removed the other messages. What about when the VC is first created?

matthew-carroll replied:
I added a `viewDidLoad` implementation and called `onUserSettingsChanged:nil` there to ensure we have the correct setting at the beginning.

stuartmorgan replied:
`viewDidAppear:` is already calling `onUserSettingsChanged:`

matthew-carroll replied:
Is there a risk of a timing issue there given that Flutter is running asynchronously? To me, `viewDidAppear` sounds like it might show the wrong brightness momentarily before Flutter updates..

stuartmorgan replied:
Given that Flutter is running asynchronously there's a risk either way; `willAppear` and `didAppear` are almost certainly in the same event loop anyway. There was a long chat thread about that originally, and I thought there was a specific plan around how to make the initial frame correct. But calling it when the engine is first created is a good start.

matthew-carroll commented:
That’s because the official docs list all those methods as places where a
VC should handle any changes to the trait collection.

On Thu, Aug 8, 2019 at 7:46 PM stuartmorgan <notifications@github.com>
wrote:

> *@stuartmorgan* commented on this pull request.
> ------------------------------
>
> In shell/platform/darwin/ios/framework/Source/FlutterViewController.mm
> <https://github.com/flutter/engine/pull/10791#discussion_r312312542>:
>
> > @@ -690,6 +695,11 @@ - (CGFloat)statusBarPadding {
>    return CGRectIsNull(intersection) ? 0.0 : intersection.size.height;
>  }
>
> +- (void)viewWillLayoutSubviews {
> +  [super viewWillLayoutSubviews];
> +  [self onUserSettingsChanged:nil];
>
> Looking more broadly at the patch rather than just the part that was
> failing before: why are all these layout-based methods getting
> onUserSettingChanged: calls? traitCollectionDidChange: makes sense, but I
> don't understand why a relayout would change any settings (and especially
> why each relayout would need to update the settings 3 times, as this change
> appears to be doing.)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/engine/pull/10791?email_source=notifications&email_token=ABXMHHHIJWI3CZO3W2V4CCLQDTK67A5CNFSM4IKPOT3KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCBCHU7A#pullrequestreview-272923260>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABXMHHEVNE3U2NEHM6TDJ43QDTK67ANCNFSM4IKPOT3A>
> .
>

matthew-carroll commented:
@chinmaygarde @stuartmorgan @dnfield @LongCatIsLooong can I merge in?

chinmaygarde commented on shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:
```diff
@@ -403,6 +403,10 @@ - (void)surfaceUpdated:(BOOL)appeared {
 
 #pragma mark - UIViewController lifecycle notifications
 
+- (void)viewDidLoad {
```
This needs to call `[super viewDidLoad]` I think.

stuartmorgan replied:
This call is unlikely to do anything where it is, because the engine is created in viewWillAppear:, which is after this. So the channel will be nil, and the message will go nowhere.

This should be right after the launchEngine: call just below (and in the initWithEngine: for that path)

matthew-carroll commented:
@stuartmorgan I moved the sending of user settings to the location you suggested.

FYI @darrenaustin @chinmaygarde @stuartmorgan I added the concept of platform contrast to this PR. We'll need it for Dark Mode, too, and it's essentially the same behavior that this PR added for platform brightness.

Please let me know if you think we're good to merge in with these changes.

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:
```diff
@@ -940,6 +950,34 @@ - (BOOL)isAlwaysUse24HourFormat {
+  if (@available(iOS 13, *)) {
+    UIAccessibilityContrast contrast = self.traitCollection.accessibilityContrast;
+
+    if (contrast == UIAccessibilityContrastHigh) {
```
This constant only exists in the iOS 13 SDK, unlike UIUserInterfaceStyleDark which is in 12. This isn't going to compile on the bots.

If you want to add this we need to go back to the early discussion of declaring the constant yourself, conditionally, behind an ifdef based on the compiling SDK, so that you can use the constant regardless of which SDK you're compiling against. Same with adding a declaration of accessibilityContrast, which is also 13+

matthew-carroll replied:
Do you have an example of defining such a constant that I can reference and replicate?

matthew-carroll replied:
I took a stab at only including the accessibility logic when the max API is 13 or above. Please let me know if that's sufficient.

stuartmorgan replied:
As previously discussed, putting the actual implementation behind ifdefs doesn't do what you want, because that would mean it would only work if someone compiles their own engine locally with the Xcode beta; none of this code would actually be shipped in our library.

If this code is important to have now, then per the previous discussion what you need are declarations of the constant and the method that are behind reversed SDK ifdefs (i.e., the local declarations are used when compiling with a stable SDK, and not there to cause duplicate definition errors when using the beta SDK), allowing all of the logic here to compile regardless of SDK.

I don't have an example offhand, and would have to dig one up tomorrow. But the declarations that you need behind the ifdefs are exactly the ones that are in the beta SDK: the same name and value for the constant, and the identical signature for the method (declared via a class extension).

matthew-carroll replied:
An example to replicate would be great. I wasn't clear on the distinction last time and I haven't been able to locate an example of it in practice.

stuartmorgan replied:
Here are some examples; they are old and on macOS, but the principle is exactly the same.
method: https://hg.mozilla.org/camino/file/d974e1a3e6f55d09feb5afdf5ee50aa97f846518/src/browser/BrowserWindowController.mm#l108
constant: https://hg.mozilla.org/camino/rev/3249ec97333c#l3.12

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.m:
```diff
@@ -23,4 +29,172 @@ - (void)testBinaryMessenger {
+
+- (void)testItReportsPlatformBrightnessWhenViewWillAppear {
+  // Setup test.
+  id engine = OCMClassMock([FlutterEngine class]);
```
Have you verified that `engine` won't outlive this test? Per offline discussion, if a class mock outlives the test, Bad Things happen. I think the recent refactors to method channel handling means that it won't leak the engine like it used to, but I'd like to be sure.

matthew-carroll replied:
How might I verify this?

matthew-carroll replied:
I just went ahead and explicitly cleaned up every mock in each test.

matthew-carroll commented:
@stuartmorgan I pushed an update that hopefully contains the correct implementation of the conditional property inclusion that we discussed today. This appears to compile and run through tests without issue locally against 12.2 and 13. Please let me know if you think we're good to merge in with this round of changes.

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:
```diff
@@ -32,6 +32,18 @@ @interface FlutterViewController () <FlutterBinaryMessenger>
+  UIAccessibilityContrastHigh = 2
+} UIAccessibilityContrast;
+
+@interface UITraitCollection (AccessibilityContrastApi)
```
Nit: I'd suggest naming the extension something like `MethodsFromNewerSDK`, rather than `AccessibilityContrastApi`, to make it very obvious that this is a declaration of something from the SDK rather than a normal extension (with an implementation).

If you do keep the current name, s/Api/API/ per ObjC naming style.

matthew-carroll replied:
done

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:
```diff
@@ -32,6 +32,18 @@ @interface FlutterViewController () <FlutterBinaryMessenger>
 @property(nonatomic, readwrite, getter=isDisplayingFlutterUI) BOOL displayingFlutterUI;
 @end
 
+#if __IPHONE_OS_VERSION_MAX_ALLOWED < 130000
```
Since it seems pretty clear from all the discussions that most of the Flutter team is not familiar with the general approach of what is being done in this block, please add a comment explaining that this is declaring things from the iOS 13 SDK so that code in this class can compile even though we're still using the iOS 12 SDK.

matthew-carroll replied:
done

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:
```diff
@@ -962,6 +984,34 @@ - (BOOL)isAlwaysUse24HourFormat {
   return [dateFormat rangeOfString:@"a"].location == NSNotFound;
 }
 
+- (NSString*)brightnessMode {
```
Google's ObjC style requires declaration comments on all non-trivial methods, including private ones. Please comment this and the next method. A good thing to explain as part of that would be what these seemingly-arbitrary-in-this-context strings are (i.e., that they are values expected by the framework for the <insert key name> key in the settings channel).

http://google.github.io/styleguide/objcguide.html#declaration-comments

matthew-carroll replied:
done

stuartmorgan commented on shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.m:
```diff
@@ -23,4 +41,200 @@ - (void)testBinaryMessenger {
+  [mockTraitCollection stopMocking];
+}
+
+- (UITraitCollection*)setupFakeTraitCollectionWithContrast:(UIAccessibilityContrast)contrast {
```
s/setupFakeTrait.../fakeTrait.../

This isn't operating on a trait collection, it's returning one, so should be named like a getter.

Also, this should get a brief declaration comment as well.

matthew-carroll replied:
done
